### PR TITLE
[9.0] [FIX] attachment_s3: Migrate to boto3

### PR DESCRIPTION
**Pull request by sebalix**
Backported https://github.com/camptocamp/odoo-cloud-platform/pull/48/commits/c61cf6c4e57fe7526c4df01ec32eaeb06023dc21

### CODE REVIEW

File: attachment_s3/models/ir_attachment.py [hunk 1]
```diff
@@ -6,10 +6,10 @@
 import base64
 import inspect
 import logging
+import StringIO
```
Comment (simahawk) [inline attachment_s3/models/ir_attachment.py]: use cStringIO if available

File: attachment_s3/models/ir_attachment.py [hunk 2]
```diff
@@ -192,31 +201,34 @@ def _get_s3_bucket(self, name=None):
                     ) % (bucket_name, bucket_name)
 
             raise exceptions.UserError(msg)
-
+        # try:
```
Comment (simahawk) [inline attachment_s3/models/ir_attachment.py]: "try" comment to be removed?

File: attachment_s3/models/ir_attachment.py [hunk 3]
```diff
@@ -227,10 +239,16 @@ def _file_read_s3(self, fname, bin_size=False):
...
-        filekey = bucket.get_key(s3uri.item())
-        if filekey:
-            read = base64.b64encode(filekey.get_contents_as_string())
-        else:
+        try:
+            key = s3uri.item()
+            bucket.meta.client.head_object(
+                Bucket=bucket.name,  Key=key
+            )
+            res = StringIO.StringIO()
```
Comment (simahawk) [inline attachment_s3/models/ir_attachment.py]: shouldn't close this?

File: attachment_s3/models/ir_attachment.py [hunk 4]
```diff
@@ -245,57 +263,55 @@ def _file_read(self, fname, bin_size=False):
...
-            filekey = bucket.get_key(key) or bucket.new_key(key)
+            obj = bucket.Object(key=key)
+            file_ = StringIO.StringIO()
+            file_.write(bin_data)
+            file_.seek(0)
             filename = 's3://%s/%s' % (bucket.name, key)
             try:
-                filekey.set_contents_from_string(bin_data)
-            except S3ResponseError as error:
+                obj.upload_fileobj(file_)
```
Comment (simahawk) [inline attachment_s3/models/ir_attachment.py:276]: same: the in-memory file should be closed once consumed

**Review by simahawk — CHANGES_REQUESTED** [review]

**Comment by simahawk** [comment]
FTR I think we should move out of our modules the handling of S3 read/write (maybe for v14).
I already added support for server env variables.
https://github.com/OCA/storage/tree/13.0/storage_backend_s3

File: attachment_s3/models/ir_attachment.py (hunk 3 above)
Comment (sebalix) [reply attachment_s3/models/ir_attachment.py]: @simahawk I tried with my last fixup but it didn't work.
Not supported in Python 2: https://bugs.python.org/issue1286

```python
[...]
  File "/odoo/external-src/odoo-cloud-platform/attachment_s3/models/ir_attachment.py", line 246, in _file_read_s3
    with StringIO() as res:
AttributeError: __exit__
```

**Comment by sebalix** [comment]
@simahawk I replaced `cStringIO` by `StringIO` and tested locally.
No context manager available however on these classes in Python 2.

File: attachment_s3/models/ir_attachment.py (hunk 3 above)
Comment (simahawk) [reply attachment_s3/models/ir_attachment.py]: you have to use `contextlib.closing` :wink: 
```
from contextlib import closing
[...]
with closing(...):
    # do it
[...]
```
or simply close it "manually" once you are done

**Comment by simahawk** [comment]
>  No context manager available however on these classes in Python 2.

https://docs.python.org/2.7/library/contextlib.html#contextlib.closing

**Comment by yvaucher** [comment]
@sebalix can you rebase or should we close it?

**Pull request closed without merging**